README.md

```diff
@@ -0,0 +1,171 @@
+openssl genrsa -out client1.key 2048
+openssl genrsa -out client2.key 2048
+
+openssl req -new -key server.key -subj "/CN=server" -out server.csr
+openssl req -new -key client1.key -subj "/CN=client1" -out client1.csr
+openssl req -new -key client2.key -subj "/CN=client2" -out client2.csr
+
+# sign CSR with CA
+openssl x509 -req -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt -days 365 -extfile host-server.conf
+openssl x509 -req -in client1.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out client1.crt -days 365
+openssl x509 -req -in client2.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out client2.crt -days 365
+```
+
+#### Resource control
+
+The control over compute, network and memory resources is provided by means of cgroups.
+Setting up the cgroups requires root access. However, running the server with root permissions is against best security practice.
+For this reason, I would suggest configuring cgroups in advance. Then the server will add itself to these cgroups at the beginning of process execution. 
+
+### Library
+
+The library is the core of the server, which does the following:
+ - implements process management APIs:
+   - `StartProcess` starts a new process.
+   - `GetProcessStatus` returns process status.
+   - `StreamOutput` returns combined process standard and error output stream.
+   - `StopProcess` stops the process.
+ - verifies user authorization:
+   - the library maintains an authorization table of clients and corresponding bitmap of permitted APIs.
+   - at the beginning of each API call the library checks that the client is authorized to call the API.
+   - for `GetProcessStatus`, `StreamOutput`, and `StopProcess` APIs, the library verifies that the process has been created by the same client.
+   - *note:* the library does not have the notion of the `superuser`, but it can be easily added.
+ - generates UID for processes. The process UID is an `uint64` integer atomically increased every time a new process is initiated.
+ - maintains a process table, for both active and terminated processes. The key is process UID. The values is a structure representing the process:
+ ```
+type Process struct {
+	clientID string    // client ID retrieved from the client certificate 
+	cmd      *exec.Cmd // process object in Go
+	logfile  string    // path to the file containing combined process standard and error outputs
+	status   string    // process status
+}
+ ```
+
+The library exposes singleton process manager object holding all aforementioned structures:
+```
+const (
+	LogDir = "/tmp/gravitest/proc/logs"
+
+	PermStart  = 0x01
+	PermStop   = 0x02
+	PermStatus = 0x04
+	PermStream = 0x08
+
+	StatusInit    = "initializing"
+	StatusRunning = "running"
+	StatusFailed  = "failed"
+	StatusStopped = "stopped"
+)
+
+type ProcManager struct {
+	// process table [process UID : Process]
+	procs     map[uint64]*Process
+	procMutex sync.Mutex
+	// permission table [client ID : permission bitmap]
+	perm map[string]int
```

Comment:
This model is not extensive. If a new user is added the mapping needs to be updated.

Also does the userA with `PermStream` and `PermStop` permissions will be able to  stop or read other user's jobs ?

Reply:
> This model is not extensive. If a new user is added the mapping needs to be updated.
Correct, my design assumes that the users are known upfront. 
I can extend the functionality of adding users (for example, adding configuration API and the notion of super-user), but AFAIK, this is out of scope of this project.  Let me know otherwise.
> Also does the userA with PermStream and PermStop permissions will be able to stop or read other user's jobs ?
No, userA can only access its own processes.

Reply:
Yes, agree. I was thinking about having a hardcoded groups/permission mapping that will allow to create many user without code modification but let's keep current version.